File: Data Structures & Algorithms/invert-a-binary-tree/submission-7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # Return None if the root is None
        if not root:
            return None

        # Swap the left and right children node
        root.left, root.right = root.right, root.left

        # Recursive call for each left and right node
        self.invertTree(root.left)
        self.invertTree(root.right)

        # Return the modified tree's root node
        return root

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-7.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        const int inorderSize = static_cast<int>(inorder.size());
        int preIdx = 0;
        unordered_map<int, int> m;

        // Iterate through inorder to create inorder's value to index map
        for (int i = 0; i < inorderSize; ++i) {
            m[inorder[i]] = i;
        }

        // subtree's range= [L, R]
        auto dfs = [&](this auto&& self, int L, int R) -> TreeNode* {
            // Return empty when L crosses R
            if (L > R) {
                return nullptr;
            }

            // Get root node's value from preorder[preorder's index]
            // and then increment preorder's index
            int rootVal = preorder[preIdx++];

            // Create root node
            TreeNode* root = new TreeNode(rootVal);

            // Get the index of root node in inorder
            int M = m[rootVal];

            // inorder = (L, M - 1) | root | (M + 1, R)
            root->left = self(L, M - 1);
            root->right = self(M + 1, R);
            return root;
        };
        return dfs(0, inorderSize - 1);
    }
};
